BAD-372: Thoth Telescope

keegansmith21 opened the pull request:
# Thoth

This is a PR for the creation of the Thoth telescope into OAEBU workflows.

This also updates the API seeding to create the Open Book Publishers organisation and the relevant workflows.

## Design

The design of this telescope varies from others, as we wanted to minimise the layers of abstraction (inheritance) and the reliance on the release class. Comments and suggestions are very welcome!

The Thoth release class holds only the release date and DAG ID (this too can be removed using dynamic airflow tasks post version upgrade #112). The rest of the information normally held in the release class has been delegated to the telescope class.

The telescope inherits from the base 'Workflow' class rather than any of the specific telescope classes. This removes one layer of inheritance in its design, making its functionality and operation more transparent. This also means that some functions that would normally be called from the specific telescope classes (ie. streamtelescope, snapshottelescope) have been recreated as standalone functions (get_data_path(), blob_name_from_path(), make_workflow_folder()). Currently these functions are stored in thoth_telescope.py, but these should be moved to a more generic utility file as more telescopes make use of them.

Recreating some of the core functions of the telescope allows for the redesign of the folder hierarchy. The updated hierarchy can be deduced from [this line](https://github.com/The-Academic-Observatory/oaebu-workflows/pull/119/files#diff-9bd08955dc1abcb10519d5f53eb247208b1fe879a5dfd5a0826866ac76985ae3R297). The result should look like this:
data_path/thoth_onix_org/thoth_onix_org_2023_01_01/download
data_path/thoth_onix_org/thoth_onix_org_2023_01_01/transform

## Workflow
The workflow for Thoth is quite simple:
1. Download publisher metadata (ONIX file)
2. Transform the metadata in much the same way as the ONIX telescope
3. Upload the ONIX feed to Bigquery

Since this is effectively an alternate way of retrieving an ONIX feed, it uses some of the ONIX labelling (including having the same dataset ID). However, the DAG ID and download files are prefixed with "thoth" to avoid confusion. The workflow also shares the same 'onix.json' schema file, as we would expect this to be no different to the onix telescope's.

keegansmith21 commented:
@jdddog just a question about the design that would be good to have your input on when you're ready to review:
This currently works as a standalone telescope in that it'll pull data from Thoth, transform it and create a BQ table. There is another option for the telescope design that I did not consider until after creating it... We could make the workflow follow a similar pattern as with the OAPEN metadata telescope (#118). That is to say, we could download the onix XML (from Thoth) and upload it to the SFTP server's upload file. Then we could simply let the ONIX telescope pick up the file and process it as normal. This would have the advantage of confining the entirety of ONIX processing to a single telescope type. Please let me know what you think :smiley:

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+    def make_release(self, **kwargs) -> ThothRelease:
+        release_date = make_release_date(**kwargs)
+        release = ThothRelease(dag_id=self.dag_id, release_date=release_date)
+        self.download_folder = make_workflow_folder(self.dag_id, release_date, SubFolder.downloaded.value)
+        self.transform_folder = make_workflow_folder(self.dag_id, release_date, SubFolder.transformed.value)
```
I would modify this a little bit, so that the folder created by `get_data_path(), dag_id, f"{dag_id}_{release_string}"` can be accessed separately, e.g. as self.workflow_folder.

Perhaps then `make_workflow_folder` could just create the path at `workflow_folder` and you can make the download_folder and transform_folder after this with os.path.join.

This is so that you can use `self.workflow_folder` to cleanup the whole workflow folder.

keegansmith21 replied:
What would you think of the following as a solution?:
```suggestion
def make_workflow_folder(dag_id: str, release_date: pendulum.DateTime, *subdirs: str) -> str:
    release_string = release_date.format("YYYY_MM_DD")
    path = os.path.join(get_data_path(), dag_id, f"{dag_id}_{release_string}", *subdirs)
    os.makedirs(path, exist_ok=True)
    return path
```

jdddog replied:
Yeah that looks good

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+    def upload_downloaded(self, release: ThothRelease, **kwargs) -> None:
+        """Upload the downloaded thoth onix XML to google cloud bucket
+
+        :raises AirflowException: Raised if there is not exactly 1 file in the donwload folder matching the expected name
```
Add the remainder of the docstrings.

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+        return release
+
+    def download(self, release: ThothRelease, **kwargs) -> None:
+        """Task to download the ONIX release from Thoth."""
```
Add the remainder of the docstrings.

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+
+    def download(self, release: ThothRelease, **kwargs) -> None:
+        """Task to download the ONIX release from Thoth."""
+        download_onix(
+            publisher_id=self.publisher_id,
+            download_folder=self.download_folder,
+            format_spec=self.format_specification,
+        )
```
```suggestion
        thoth_download_onix(
            host_name=self.host_name,
            publisher_id=self.publisher_id,
            download_folder=self.download_folder,
            format_spec=self.format_specification,
        )
```

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+
+    def transform(self, release: ThothRelease, **kwargs) -> None:
+        """Task to transform the Thoth ONIX data"""
+        parse_onix(self.download_folder, self.transform_folder, make_org_id(self.organisation_name))
```
The source variable of `parse_onix` is used to create a field called `COKI_ID`, which we don't use anywhere. We can remove the need for `make_org_id(self.organisation_name)` by removing this from the ONIX Parser.

These are the lines to remove in ONIX Parser, along with the associated `data_src` variables:
https://github.com/The-Academic-Observatory/onix-parser/blob/main/src/main/java/academy/observatory/app/OnixParser.java#L208-L209

keegansmith21 replied:
Sure! I've create a [PR](https://github.com/The-Academic-Observatory/onix-parser/pull/7) for the ONIX Parser

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+        )
+
+    def upload_transformed(self, release: ThothRelease, **kwargs) -> None:
+        """Upload the downloaded thoth onix .jsonl to google cloud bucket"""
```
Add remaining docstrings.

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+        )
+
+    def cleanup(self, release: ThothRelease, **kwargs) -> None:
+        """Delete all files and folders associated with this release."""
+        for path in [self.download_folder, self.transform_folder]:
+            try:
+                shutil.rmtree(path)
+            except FileNotFoundError as e:
+                logging.warning(f"No such file or directory {path}: {e}")
```
It would be useful to move the cleanup functionality into a reusable function, as this will be reused across workflows, see below for function. Typically the cleanup should remove the workflow folder and also remove any XComs.

```suggestion
        """Delete all files and folders associated with this release."""
        
        cleanup(self.workflow_folder, self.dag_id, kwargs["execution_date"])
```

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+                shutil.rmtree(path)
+            except FileNotFoundError as e:
+                logging.warning(f"No such file or directory {path}: {e}")
+
```
We can remove the entire workflow_folder rather than each subfolder.

```suggestion
def cleanup(workflow_folder, dag_id: str, execution_date):
    # Cleanup workflow folder
    try:
        shutil.rmtree(workflow_folder)
    except FileNotFoundError as e:
        logging.warning(f"No such file or directory {workflow_folder}: {e}")

    # Cleanup XComs
    delete_old_xcoms(dag_id=dag_id, execution_date=execution_date)
```

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+
+
+def download_onix(
+    publisher_id: str,
```
If hostname can be altered, then better to let it be specified with an argument.

```suggestion
    *,
    host_name: str,
```

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+) -> None:
+    """Hits the Thoth API and requests the ONIX feed for a particular publisher.
+    Creates a file called onix.xml at the specified location
+
```
```suggestion

    :param host_name: the Thoth hostname.
```

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+    :param format_spec: The ONIX format to use. Options can be found with the /formats endpoint of the API
+    :param num_retries: The number of times to retry the download, given an unsuccessful return code
+    """
+    url = (
+        THOTH_BASE_URL
+        + THOTH_FORMAT_EXTENSION.format(format_specification=format_spec)
+        + THOTH_PUBLISHER_EXTENSION.format(publisher_id=publisher_id)
+    )
```
```suggestion
    url = THOTH_URL.format(host_name=host_name, format_specification=format_spec, publisher_id=publisher_id)
```

jdddog commented on oaebu_workflows/workflows/thoth_telescope.py:
```diff
@@ -0,0 +1,331 @@
+    """
+    release_string = release_date.format("YYYY_MM_DD")
+    path = os.path.join(get_data_path(), dag_id, f"{dag_id}_{release_string}", folder_name)
+    if not os.path.exists(path):
```
I probably put this line before `if not os.path.exists(path):` but it is not necessary.

```suggestion
```

jdddog reviewed (changes requested):
Hey Keegan, the workflow looks good, it is much easier to understand having less inheritance. 

> This currently works as a standalone telescope in that it'll pull data from Thoth, transform it and create a BQ table. There is another option for the telescope design that I did not consider until after creating it... We could make the workflow follow a similar pattern as with the OAPEN metadata telescope (#118).

I prefer how you have made the workflow here, because it is a more straightforward way to get the data into BigQuery.

The main changes suggested are:
* `publisher_id` can be a parameter passed to the workflow and stored in the extra field of the workflow in the API.
* It would be good to remove `organisation_name`. It is inconsistent that this parameter
is used to create the `dataset_description` inside the workflow, and the `dag_id` from outside the workflow
in the dag file. I think it would be cleaner to remove it
entirely and set both the `dag_id` and `dataset_description` in the dag file. We can remove the need for the organisation name for `parse_onix`.
* By having a separate variable for the `workflow_folder`, we can make the cleanup simpler, and also put the cleanup functionality into a function.

keegansmith21 commented:
@jdddog I've implemented the changes you've requested. Apart from those, I also made these changes:

- Added test for the parse_onix() function in the onix_telescope
- Removed date_format from onix_telescope and API seed as it does not appear to be used anywhere

jdddog reviewed (approved):
Hey Keegan, the changes look really good!
